Issue 3093 - Download rules JSON for content blocker app

dedecej opened the pull request:
What has been done:
- Refresh button in setting menu
- Login for downloading and updating rules
- Allowed background executions for reloading rules in Safari
- Enabled background fetching

dedecej committed:
Added new video.

dedecej committed:
Small fixes.

dedecej committed:
Added activity indicator

dedecej committed:
Add Eyeo's licence agreements.

Commit:
Switched from com.salsitasoft to org.adblockplus, fixed AppGroup entitlement

dedecej committed:
Merge pull request #1 from adblockplus/master

Merge with Eyeo upstream

dedecej committed:
Merge pull request #2 from adblockplus/master

Sync with Eyeo

dedecej committed:
Merge branch 'master' into develop

dedecej committed:
Added update button and filterlists updating logic.

Change-Id: 26a9da5399
Story-Id: 1273354/stories/102908894

dedecej committed:
Implemented background fetching.

Change-Id: a0bc344729
Story-Id: 1273354/stories/102908894

fhd commented on AdblockPlusSafari/AdblockPlus.m:
```diff
@@ -17,15 +17,15 @@
+NSString *AdblockPlusActivated = @"AdblockPlusActivated";
 static NSString *AdblockPlusEnabled = @"AdblockPlusEnabled";
 static NSString *AdblockPlusAcceptableAdsEnabled = @"AdblockPlusAcceptableAdsEnabled";
+static NSString *AdblockPlusFilterlists = @"AdblockPlusFilterlists";
```
Nit: "filter list" being two words, the name here should be `AdblockPlusFilterLists` with an upper case L. There are more instances of it being written as a single word in the pull request that should be updated.

dedecej replied:
I see, I copied this mistake from invisio designs. Unfortunately, it is scattered across multiple places.

fhd commented on AdblockPlusSafari/AdblockPlus.m:
```diff
@@ -36,16 +36,25 @@ - (instancetype)init
+        AdblockPlusDownloadedVersion: @1,
+        AdblockPlusFilterlists:
+          @{@"https://easylist-downloads.adblockplus.org/easylist_content_blocker.json":
+              @{@"filename": @"easylist"},
```
The file names on the disk should also be changed accordingly, see https://issues.adblockplus.org/ticket/3093.

dedecej replied:
OK

fhd commented on AdblockPlusSafari/AdblockPlusController.m:
```diff
@@ -100,6 +115,20 @@ -(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NS
+    dateFormatter.locale = [NSLocale currentLocale];
+    dateFormatter.dateStyle = NSDateFormatterFullStyle;
+    dateFormatter.timeStyle = NSDateFormatterShortStyle;
+    return [NSString stringWithFormat:@"Last filterlist update: %@", [dateFormatter stringFromDate:lastUpdate]];
```
Same as with the variable name, "filter list" should be two words - that was wrong in the mockup.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.h:
```diff
@@ -0,0 +1,32 @@
+//
```
Should be our official license header, likewise for the other files below.

dedecej replied:
Yeah, I forgot.

fhd commented on AdblockPlusSafari/AdblockPlus.h:
```diff
@@ -17,22 +17,28 @@
-- (void)setEnabled:(BOOL)enabled reload:(BOOL)reload;
-
-- (void)setAcceptableAdsEnabled:(BOOL)enabled reload:(BOOL)reload;
+@property (nonatomic) NSInteger installedVersion;
```
Maybe I'm missing something, but I get the impression we could reduce complexity here and achieve the same effect by replacing the `installedVersion` and `downloadedVersion` counters with a single boolean `dirty` flag (actually not sure about the best name, `refreshed` or `applied` also come to mind).

dedecej replied:
You are right. Actually, it seemed to me as good idea to keep version of filter list, because I known how many times it was updated. For the end user product it is not really useful. I introduced dirty property.

dedecej replied:
I talked to Pavel about it and we decided to stay with version. The reason behind this is some case dirty flag might be overridden with newest update filter. Then it would take 5 days to update filter.

fhd replied:
Fair enough.

fhd commented on AdblockPlusSafari/AdblockPlus.m:
```diff
@@ -73,53 +82,42 @@ - (void)setActivated:(BOOL)activated
-#pragma mark -
-
-- (NSString *)contentBlockerIdentifier
+- (void)setFilterlists:(NSDictionary<NSString *,NSDictionary<NSString *,NSObject *> *> *)filterlists
```
I guess it's my lack of Objective-C knowledge, but I get the impression this setter is not actually used anywhere. Is it? If it is indeed not used, I would prefer to get rid of it - I'm not a fan of unused code.

dedecej replied:
Actually, it is used. Compiler will replace assignments self.filterlists = ...; to [self setFilterlists:...]; for you. It is just syntactical sugar. 

Objective-C is quite old and obsolete language. That is why I would prefer to write content blocker in swift.

fhd replied:
I see, fine then. Well, like I said - I'd go nuts if I have to add another language to the mix of those I need to deal with right now, and we have to use Obj-C anyway for other reasons. But in the long run we can probably move to Swift.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,269 @@
+
+    NSString *errorMessage;
+
+    {
```
I think this block should be a separate function. That would make the gotos unnecessary - the function could simply return instead.

fhd commented on AdblockPlusSafari/AppDelegate.m:
```diff
@@ -20,14 +20,23 @@
 #import "Appearence.h"
 #import "RootController.h"
 
+const NSTimeInterval FilterlistsUpdatePeriod = 3600*24;
```
We should go with 5 days here, for now. EasyList itself has a mechanism for reducing the download interval in case there is just too much traffic - here we don't have that.

dedecej replied:
OK

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,269 @@
+      NSString *title = @"Updating filterlists";
+      NSString *message = [NSString stringWithFormat:@"Filterlist %@ cannot be updated.%@", filterlistName, errorMessage];
+
+      UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
```
Just to make sure - if a download fails, it isn't re-attempted, right? We have some logic to ensure retries don't cause a cascade bringing down the entire filter list download infrastructure in all of our clients, but here I'd prefer to just not attempt any immediate retries for now.

dedecej replied:
Correct, it is not re-attempted. It will try it again with next refresh. The problem is that iOS allow apps execute in background for just limited time.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,269 @@
+  }
+}
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
```
So, this is invoked with an interval of `UIApplicationBackgroundFetchIntervalMinimum`, which is presumably 10 minutes? Is there an initial delay, before the first invocation? If possible, we should delay the first fetch by one minute to keep broken clients from causing traffic.

dedecej replied:
I agree, It does not make sense, I did mistake by setting interval so low. What about fetching interval once or twice per day?

fhd replied:
We usually do it every hour, but I actually think 10 minutes is fine, if iOS lets us do that... Just a delay before the first fetch would be nice.

fhd replied:
Can we do the one minute delay?

fhd replied:
Actually, let's scratch the one minute thing, it doesn't matter here. See my comment below about it. Let's leave it as it is, I'll update the issue.

fhd commented:
Alright, big sorry for the delay, I've finished reviewing now, but I didn't go too much into details. Some things stood out that we had to specify, I've updated the issue accordingly. Also noticed some smaller things, but looks good all in all already.

dedecej committed:
Merge branch 'story/download-rules' into develop

dedecej committed:
Added salsaflow config

Change-Id: 1d8c5b9ba6

dedecej committed:
Review comments.

Change-Id: f5fe433522
Story-Id: 1431320/stories/102908894

dedecej committed:
Update salsaflow

Change-Id: 79c068af61
Story-Id: 1431320/stories/102908894

dedecej committed:
File updates

Change-Id: 32be799417
Story-Id: 1431320/stories/102908894

Commit:
Implemented poor man's almost-markdown renderer

Change-Id: 90322bbed0

Commit:
Modified storyboard and project

Change-Id: e6274baca1

Commit:
Initial translation resource and screenshots

Change-Id: 01538b3ef3

Commit:
Additional text improvements https://www.pivotaltracker.com/story/show/104974740

Change-Id: 474c8d7e59

Commit:
New translation screenshots

Change-Id: 0df28c7b3f

Commit:
Fixed hardcoded label string to be localizable

Change-Id: 2c9f5529f5

Commit:
New label string localization export

Change-Id: 0fad4b8b60

fhd commented on .salsaflow/config.yml:
```diff
@@ -0,0 +1,40 @@
+# This field keeps the timestamp of when 'repo bootstrap' was executed.
```
I left a comment about this one or two days ago, but I have actually no idea where it went :suspect: 

So for the record: Please leave `.salsaflow` out of the upstream repository (i.e. this one), it doesn't make sense to anyone outside Salsita. Should suffice if you keep it in your fork.

dedecej committed:
Updated salsaflow

Change-Id: 614cc86585

dedecej committed:
Fixed issue with first update and refactored error dialog handling.

hange-Id: 8ec8bbff29
Story-Id: 1431320/stories/104997472
Change-Id: 963f3935b6

dedecej committed:
Fixed conditions.

Change-Id: 9bf547c98d
Story-Id: 1431320/stories/104997472

dedecej committed:
Remove salsaflow for pull request

Change-Id: eaf0ac4df2

dedecej commented:
Newest changes:
- Initial support for localisation
- Filter lists updated is started right after first startup of the application.
- Refactored error message handling. The error message is display only once for download failure of both filter list. The displaying is postponed until user has not activated the application. Successful automatic update will deny from displaying of scheduled error message.
- Filter list links were moved from code to separate file.

The reloading functionality has been tested and now it is ready for release.

dedecej commented on AdblockPlusSafari/NSString+MarkdownRenderer.m:
```diff
@@ -0,0 +1,48 @@
+
+- (NSAttributedString*)markdownSpanMarkerChar:(NSString*)markerChar
+                                 renderAsFont:(UIFont*)font {
+  NSError* err = nil;
```
Pavel's implementation of markdown parser (support for localisation, bold fonts).

dedecej commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+    return;
+  }
+
+  if (UIApplication.sharedApplication.applicationState != UIApplicationStateActive) {
```
App has to be in foreground

dedecej commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+  UIViewController *viewController = UIApplication.sharedApplication.delegate.window.rootViewController;
+  if ([viewController isKindOfClass:[RootController class]]) {
+    RootController *rootController = (RootController *)viewController;
+    if (![rootController.viewControllers.firstObject isKindOfClass:[HomeController class]]) {
```
HomeController need to be loaded in hierarchy (message will not display in welcome guide).

dedecej commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+
+    NSMutableDictionary *mutableFilterList = [filterlist mutableCopy];
+    mutableFilterList[@"updating"] = @NO;
+    mutableFilterList[@"lastUpdateFailed"] = @YES;
```
If execution of code get here then updating of filter list has failed.

dedecej commented on AdblockPlusSafari/AppDelegate.m:
```diff
@@ -36,34 +49,154 @@ - (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:
 {
   [self.adblockPlus checkActivatedFlag];
+
+  if (!self.firstUpdateTriggered && !self.adblockPlus.updating && self.adblockPlus.lastUpdate == nil) {
```
Our QA was complaining, that filter list is not updated right after the first start of the application. It make sense, since filter lists might be outdated for months.

self.adblockPlus.lastUpdate == nil ... It means that there any of possible previous update haven't been successful.

fhd replied:
Yeah, they should be updated on startup. Note that we should _not_ wait for one minute in this case. Since only the first startup really matters for the scheduling how I understand it, we can forget about the one minute thing altogether. I'll update the issue, and my comment above.

dedecej commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+  }
+}
+
+- (void)updateFilterlists:(BOOL)userTriggered
```
userTriggered flag is used just for different error messages. It does not have more use.

dedecej commented:
@fhd: Pull request is ready for review.

fhd commented on Translations/en.xliff:
```diff
@@ -0,0 +1,140 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Can we move the translation related changes to a separate pull request? I'd like to have only issue 3093 tackled here.

fhd commented on AdblockPlusSafari/AdblockPlusContainerController.m:
```diff
@@ -40,22 +41,9 @@ - (void)viewDidLoad
 
   // This is the simpliest way to print Browser word using custom bold font.
   CGFloat fontSize = self.adblockBrowserLabel.font.pointSize;
-  UIFont *lightFont = [Appearence defaultLightFontOfSize:fontSize];
```
These changes are for translations as well, right? Like I said below, we should focus on one thing in this pull request - issue 3093.

fhd replied:
It really seems like this is related to translations - as I said before, we should stick to #3093 in this pull request, it's big enough as it is.

fhd replied:
@dedecej did you notice these comments? They haven't been addressed. I _really_ want to stick to solving one issue per PR without unrelated changes.

fhd commented on AdblockPlusSafari/AppDelegate.m:
```diff
@@ -20,14 +20,27 @@
+// Update filter list every 5 days
+const NSTimeInterval FilterlistsUpdatePeriod = 3600*24*5;
+// Wake up application every 12 hours (just hint for iOS)
+const NSTimeInterval BackgroundFetchInterval = 3600*12;
```
Every 12 hours starting when? The first start of the app? Is it technically possible that the time between checks will be more than 12 hours? Does starting the app influence this interval?

It might be better to go for a shorter interval here, we use one hour in the ABP core code.

Another question: Will a fetch happen when the app is started for the first time? That should happen.

dedecej replied:
It seems to me that this value is completely ignored by iOS. On my phone it wake up app every six hours regardless value I set here.

dedecej replied:
Yes, there is an update, which executed after first start of the application.

fhd commented:
Had a look (but ignored the translation related changes). Looks good, just a few comments.

dedecej committed:
Revert "New label string localization export"

This reverts commit 6d3c6f15ee32d43706d586ce165a2872a6f2e6ec.

dedecej committed:
Revert "Additional text improvements https://www.pivotaltracker.com/story/show/104974740"

This reverts commit e512267b6b2552268a28a30f67e2a258cc841bc1.

dedecej committed:
Revert "Initial translation resource and screenshots"

This reverts commit 2c060b395ad7b7ebffe2c9634eb46edf22f1db38.

dedecej committed:
Project file fix.

dedecej committed:
Changed interval of background refresh.

dedecej commented:
We work little bit differently. I implement this feature and then Pavel created localisation. Vojta raised bug and I fixed it on current head with Pavel localisation. I tried to revert as much patches as possible but one which involves changes in storyboard is hard to revert.

dedecej commented:
@fhd ready for review.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+    viewController = viewController.presentedViewController;
+  }
+
+  NSString *title = @"​Filter lists Updating​";
```
Did Sven ask for this dialog? We usually don't show any error message when an automatic download failed. For a user triggered one we can do it, I'll add that to the issue.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+        }
+
+        // Remove filter lists whose tasks have been planned again
+        for (NSString *filterlistName in self.downloadTasks) {
```
Nit: Should be `filterListName` I guess.

fhd commented on AdblockPlusSafari/AdblockPlusExtras.m:
```diff
@@ -0,0 +1,326 @@
+
+        // Set updating flag to false of filterlist, which was cancelled by user (user killed application).
+        if ([set count] > 0) {
+          NSMutableDictionary *filterlists = [sSelf.filterLists mutableCopy];
```
Nit: `filterLists` I guess, same below and in the comment above.

fhd commented on AdblockPlusSafariExtension/AdblockPlus+Extension.m:
```diff
@@ -0,0 +1,56 @@
+
+@implementation AdblockPlus (Extension)
+
+- (NSURL *)currentFilterlistURL
```
Nit: Another "filterlist" -> "filter list" issue. I recommend a case sensitive search :smile:

dedecej committed:
Fixed wrong capitalisation of variable and changed error message in dialog.

dedecej committed:
Fixed wrong capitalisation of variable.

dedecej commented:
@fhd I hope this is final iteration of our review.

I fixed all typos in filter list variables and I updated error dialog.

I read this in issue description:
_If the download fails, it should not be retried - we should wait another full five days for the next try_
Do you really want to do next try after 5 days?

fhd commented:
> Do you really want to do next try after 5 days?

@dedecej Yeah, the problem is that we need a slightly complicated retry logic, otherwise we risk amplifying infrastructure issues. I'd like to tackle that in a follow-up issue however, this patch is big enough already.

fhd commented:
Superseded by https://github.com/adblockplus/adblockplussafariios/pull/6.
